core/domain/stats_jobs_one_off.py

```diff
@@ -16,16 +16,212 @@
+                yield (item.exploration_id,
+                       {
+                           'type': 'complete_exploration',
+                           'version': item.exploration_version,
+                           'state': item.state_name
+                       })
+
+
+    @staticmethod
+    def reduce(key, values):
+        values = map(ast.literal_eval, values)
+        values = sorted(values, key=lambda x: x['version'])
+
+        for version, events in itertools.groupby(values,
+                                                 key=lambda x: x['version']):
+            if version == 1:
+                # Get a copy of the uncorrupted *_v1 statistics
+                old_stats = stats_services.get_exploration_stats_by_id(
+                    key, version)
+                for state_stats in old_stats.state_stats_mapping.values():
+                    state_stats.total_answers_count_v2 = 0
+                    state_stats.useful_feedback_count_v2 = 0
+                    state_stats.total_hit_count_v2 = 0
+                    state_stats.first_hit_count_v2 = 0
+                    state_stats.num_times_solution_viewed_v2 = 0
+                    state_stats.num_completions_v2 = 0
+                RecomputeStatistics._recompute_statistics(
+                    key, version, events, old_stats)
+            else:
+                change_list = exp_models.ExplorationCommitLogEntryModel.get(
+                    'exploration-%s-%s' % (key, version)).commit_cmds
+
+                old_exp_version = version - 1
+                exploration_stats = stats_services.get_exploration_stats_by_id(
+                    key, old_exp_version)
+                if exploration_stats is None:
+                    raise ValueError('Recomputation of exploration with id'
+                                     '%s for version %s failed'
+                                     % (key, old_exp_version))
+                exploration_stats = (
+                    RecomputeStatistics._update_state_stats_mapping(
+                        exploration_stats, change_list))
+
+                RecomputeStatistics._recompute_statistics(
+                    key, version, events, exploration_stats)
+
+    @classmethod
+    def _update_state_stats_mapping(cls, old_stats_model, change_list):
+        """Update the state_stats_mapping to correspond with the changes
+        in change_list.
+
+        Args:
+            old_stats_model: ExplorationStatsModel. The stats model
+                for the previous verion.
+            change_list: list(dict). A list of all of the commit cmds to
+                the old_stats_model up to the next version.
+
+        Returns:
+            ExplorationStatsModel. An ExplorationStatsModel with updated
+                state_stats_mapping and version.
+        """
+        # Handling state additions, deletions and renames.
+        for change_dict in change_list:
```

Comment:
You'll need to handle exploration versions everts too.

Reply:
If we decide to use the method in stats_services this will be taken care of.